File: go/policy.go
package guardrail

import (
	"embed"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

//go:embed policies/*.json
var bundledPacks embed.FS

const defaultPack = "standard-v1"

// Bands are the probabilities at which a category resolves to each action.
type Bands struct {
	Block  float64 `json:"block"`
	Review float64 `json:"review"`
	Flag   float64 `json:"flag"`
}

// Category is one hazard in the taxonomy.
type Category struct {
	ID                   string
	Name                 string
	Description          string
	Refs                 []string
	Surfaces             map[Surface]bool
	Weight               float64
	BaseSeverity         int
	Sentinel             bool
	SentinelInstructions string
	Thresholds           map[string]Bands
	Route                Route
	NeverBelow           Action
	Enabled              bool
}

// Threshold returns the bands for a surface, falling back to "default".
func (c *Category) Threshold(surface Surface) (Bands, bool) {
	if b, ok := c.Thresholds[string(surface)]; ok {
		return b, true
	}
	b, ok := c.Thresholds["default"]
	return b, ok
}

// Applies reports whether the category is enabled and scored on a surface.
func (c *Category) Applies(surface Surface) bool {
	return c.Enabled && c.Surfaces[surface]
}

// Signal is a question asked alongside the hazard choice, read by the policy's rules.
type Signal struct {
	Name            string
	Type            string
	Surfaces        []string
	Instructions    string
	RequiresContext bool
	CriteriaRef     string
	// Criteria is kept as raw JSON so that label order survives into the request.
	Criteria    json.RawMessage
	hasCriteria bool
}

// Rule adjusts a verdict when a signal crosses a value.
type Rule struct {
	ID               string
	Signal           string
	Op               string
	Value            any
	ExceptCategories map[string]bool
	Upgrade          int
	Downgrade        int
	CapAction        Action
	FloorAction      Action
	AddFinding       string
}

// Policy is a parsed policy pack.
//
// A policy pack is a JSON document: the hazard taxonomy, the thresholds that turn a probability
// into an action, and the rules that adjust it. The descriptions in the pack are also the text
// sent to Jev as question criteria, so editing the pack changes both what the model is asked and
// how the answer is judged.
type Policy struct {
	ID          string
	Version     string
	Name        string
	ContentNote string
	Defaults    map[string]any
	Signals     map[string]*Signal
	Rules       []Rule
	Routes      map[string]string
	Categories  map[string]*Category
	// Scales are the pack's "*_scale" entries, kept raw so label order survives.
	Scales map[string]json.RawMessage

	raw map[string]json.RawMessage
}

// ParsePolicy reads a policy pack from JSON.
func ParsePolicy(data []byte) (*Policy, error) {
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return nil, fmt.Errorf("policy pack is not a JSON object: %w", err)
	}
	return fromRaw(raw)
}

// NewPolicy builds a policy from a decoded pack, such as one assembled in code.
func NewPolicy(data map[string]any) (*Policy, error) {
	encoded, err := json.Marshal(data)
	if err != nil {
		return nil, err
	}
	return ParsePolicy(encoded)
}

// LoadPolicy loads the bundled pack (""), a bundled pack by name, or a path to a JSON file.
func LoadPolicy(source string) (*Policy, error) {
	if source == "" {
		return BundledPolicy(defaultPack)
	}
	ext := strings.ToLower(filepath.Ext(source))
	if !strings.ContainsAny(source, `/\`) && ext != ".json" && ext != ".yaml" && ext != ".yml" {
		return BundledPolicy(source)
	}
	if ext == ".yaml" || ext == ".yml" {
		// Staying dependency-free matters more here than reading YAML directly.
		return nil, fmt.Errorf("YAML policy packs are not supported by the Go package; convert %s to JSON", source)
	}
	data, err := os.ReadFile(source)
	if err != nil {
		return nil, err
	}
	return ParsePolicy(data)
}

// BundledPolicy loads a pack shipped inside the package.
func BundledPolicy(name string) (*Policy, error) {
	data, err := bundledPacks.ReadFile("policies/" + name + ".json")
	if err != nil {
		return nil, fmt.Errorf("no bundled policy pack named %q", name)
	}
	return ParsePolicy(data)
}

// MustBundledPolicy is BundledPolicy for the standard pack, which is embedded and always valid.
func MustBundledPolicy() *Policy {
	p, err := BundledPolicy(defaultPack)
	if err != nil {
		panic(err)
	}
	return p
}

// QualifiedID is "id@version", the form carried on every verdict and in every cache key.
func (p *Policy) QualifiedID() string { return p.ID + "@" + p.Version }

// ForSurface lists enabled categories that apply to a surface, most serious first.
func (p *Policy) ForSurface(surface Surface) []*Category {
	var cats []*Category
	for _, c := range p.Categories {
		if c.Applies(surface) {
			cats = append(cats, c)
		}
	}
	sort.Slice(cats, func(i, j int) bool {
		a, b := cats[i], cats[j]
		if a.Weight != b.Weight {
			return a.Weight > b.Weight
		}
		if a.BaseSeverity != b.BaseSeverity {
			return a.BaseSeverity > b.BaseSeverity
		}
		return a.ID < b.ID
	})
	return cats
}

// Sentinels lists the categories that get a dedicated yes/no question on this surface.
//
// A choice question picks one label. Content can carry more than one hazard at once, and the
// ones where a miss is unacceptable get their own independent question.
func (p *Policy) Sentinels(surface Surface) []*Category {
	var out []*Category
	for _, c := range p.ForSurface(surface) {
		if c.Sentinel && c.SentinelInstructions != "" {
			out = append(out, c)
		}
	}
	return out
}

// SignalsFor lists the signal definitions that apply to a surface.
func (p *Policy) SignalsFor(surface Surface, hasContext bool) map[string]*Signal {
	out := map[string]*Signal{}
	for name, spec := range p.Signals {
		applies := false
		for _, s := range spec.Surfaces {
			if s == string(surface) {
				applies = true
				break
			}
		}
		if !applies || (spec.RequiresContext && !hasContext) {
			continue
		}
		out[name] = spec
	}
	return out
}

// CriteriaFor resolves a signal's criteria, following criteria_ref into the pack's scales.
func (p *Policy) CriteriaFor(spec *Signal) json.RawMessage {
	if spec.hasCriteria {
		return spec.Criteria
	}
	if spec.CriteriaRef != "" {
		return p.Scales[spec.CriteriaRef]
	}
	return nil
}

// MinConfidence is the confidence below which a verdict escalates toward review.
func (p *Policy) MinConfidence() float64 { return floatOr(p.Defaults["min_confidence"], 0.65) }

// OnLowConfidence is what to do with a low-confidence answer; "escalate" is the default.
func (p *Policy) OnLowConfidence() string {
	if s, ok := p.Defaults["on_low_confidence"]; ok && s != nil {
		return pyStr(s)
	}
	return "escalate"
}

// FailClosed reports whether an unreachable Jev blocks on this surface.
//
// Set per surface, because the two are not the same risk: the input check sits in front of a
// model that has its own safety, so failing open there degrades gracefully, while the output
// check is the last line and has nothing behind it.
func (p *Policy) FailClosed(surface Surface) bool {
	setting, ok := p.Defaults["on_error"]
	if !ok || setting == nil {
		return true
	}
	if perSurface, isMap := setting.(map[string]any); isMap {
		value, found := perSurface[string(surface)]
		if !found {
			return true
		}
		setting = value
	}
	return pyStr(setting) == "fail_closed"
}

// ErrorAction is the action a fail-closed surface takes when Jev is unreachable.
func (p *Policy) ErrorAction() Action {
	if s, ok := p.Defaults["error_action"].(string); ok {
		return Action(s)
	}
	return Review
}

// JSON returns the pack as JSON, as it was loaded.
func (p *Policy) JSON() ([]byte, error) { return json.Marshal(p.raw) }

func (p *Policy) String() string {
	return fmt.Sprintf("<Policy %s v%s: %d categories>", p.ID, p.Version, len(p.Categories))
}

// -- parsing --------------------------------------------------------------

func fromRaw(raw map[string]json.RawMessage) (*Policy, error) {
	var top map[string]any
	encoded, _ := json.Marshal(raw)
	if err := json.Unmarshal(encoded, &top); err != nil {
		return nil, err
	}

	p := &Policy{
		ID:          stringOr(top["id"], "custom"),
		Version:     stringOr(top["version"], "0"),
		ContentNote: stringOr(top["content_note"], ""),
		Defaults:    mapOf(top["defaults"]),
		Signals:     map[string]*Signal{},
		Routes:      map[string]string{},
		Categories:  map[string]*Category{},
		Scales:      map[string]json.RawMessage{},
		raw:         raw,
	}
	p.Name = stringOr(top["name"], p.ID)
	for k, v := range mapOf(top["routes"]) {
		p.Routes[k] = pyStr(v)
	}
	for key, value := range raw {
		if strings.HasSuffix(key, "_scale") {
			p.Scales[key] = value
		}
	}

	var sections struct {
		Categories map[string]json.RawMessage `json:"categories"`
		Signals    map[string]json.RawMessage `json:"signals"`
	}
	if err := json.Unmarshal(encoded, &sections); err != nil {
		return nil, fmt.Errorf("policy pack: %w", err)
	}
	for id, body := range sections.Categories {
		var m map[string]any
		if err := json.Unmarshal(body, &m); err != nil {
			return nil, fmt.Errorf("category %q: %w", id, err)
		}
		cat, err := parseCategory(id, m)
		if err != nil {
			return nil, err
		}
		p.Categories[id] = cat
	}
	for name, body := range sections.Signals {
		sig, err := parseSignal(name, body)
		if err != nil {
			return nil, err
		}
		p.Signals[name] = sig
	}
	if rules, ok := top["rules"].([]any); ok {
		for _, r := range rules {
			p.Rules = append(p.Rules, parseRule(mapOf(r)))
		}
	}
	return p, validate(p)
}

func parseCategory(id string, m map[string]any) (*Category, error) {
	c := &Category{
		ID:                   id,
		Name:                 stringOr(m["name"], id),
		Description:          stringOr(m["description"], ""),
		Refs:                 stringsOf(m["refs"]),
		Surfaces:             map[Surface]bool{},
		Weight:               floatOr(m["weight"], 0.5),
		BaseSeverity:         intOr(m["base_severity"], 2),
		Sentinel:             m["sentinel"] == true,
		SentinelInstructions: stringOr(m["sentinel_instructions"], ""),
		Thresholds:           map[string]Bands{},
		Route:                Route(stringOr(m["route"], "")),
		NeverBelow:           Action(stringOr(m["never_below"], "")),
		Enabled:              m["enabled"] != false,
	}
	surfaces := stringsOf(m["surfaces"])
	if len(surfaces) == 0 {
		surfaces = []string{"input", "output", "conversation"}
	}
	for _, s := range surfaces {
		c.Surfaces[Surface(s)] = true
	}
	for surface, v := range mapOf(m["thresholds"]) {
		bands := mapOf(v)
		var missing []string
		for _, name := range []string{"block", "flag", "review"} {
			if _, ok := toFloat(bands[name]); !ok {
				missing = append(missing, name)
			}
		}
		if len(missing) > 0 {
			return nil, fmt.Errorf("category %q thresholds[%s] missing %v", id, surface, missing)
		}
		c.Thresholds[surface] = Bands{
			Block:  floatOr(bands["block"], 0),
			Review: floatOr(bands["review"], 0),
			Flag:   floatOr(bands["flag"], 0),
		}
	}
	return c, nil
}

func parseSignal(name string, body json.RawMessage) (*Signal, error) {
	var s struct {
		Type            string          `json:"type"`
		Surfaces        []string        `json:"surfaces"`
		Instructions    string          `json:"instructions"`
		RequiresContext bool            `json:"requires_context"`
		CriteriaRef     string          `json:"criteria_ref"`
		Criteria        json.RawMessage `json:"criteria"`
	}
	if err := json.Unmarshal(body, &s); err != nil {
		return nil, fmt.Errorf("signal %q: %w", name, err)
	}
	var probe map[string]json.RawMessage
	_ = json.Unmarshal(body, &probe)
	_, has := probe["criteria"]
	return &Signal{
		Name:            name,
		Type:            s.Type,
		Surfaces:        s.Surfaces,
		Instructions:    s.Instructions,
		RequiresContext: s.RequiresContext,
		CriteriaRef:     s.CriteriaRef,
		Criteria:        s.Criteria,
		hasCriteria:     has,
	}, nil
}

func parseRule(m map[string]any) Rule {
	when, then := mapOf(m["when"]), mapOf(m["then"])
	r := Rule{
		ID:               stringOr(m["id"], "rule"),
		Signal:           stringOr(when["signal"], ""),
		Op:               stringOr(when["op"], "=="),
		Value:            when["value"],
		ExceptCategories: map[string]bool{},
		Upgrade:          intOr(then["upgrade"], 0),
		Downgrade:        intOr(then["downgrade"], 0),
		CapAction:        Action(stringOr(then["cap_action"], "")),
		FloorAction:      Action(stringOr(then["floor_action"], "")),
		AddFinding:       stringOr(then["add_finding"], ""),
	}
	for _, c := range stringsOf(m["except_categories"]) {
		r.ExceptCategories[c] = true
	}
	return r
}

func validate(p *Policy) error {
	if len(p.Categories) == 0 {
		return fmt.Errorf("policy pack has no categories")
	}
	for id, cat := range p.Categories {
		_, hasInput := cat.Threshold(SurfaceInput)
		if _, hasDefault := cat.Thresholds["default"]; !hasInput && !hasDefault {
			return fmt.Errorf("category %q has no default thresholds", id)
		}
		for surface, b := range cat.Thresholds {
			if !(b.Block >= b.Review && b.Review >= b.Flag) {
				return fmt.Errorf("category %q thresholds[%s] are not ordered block >= review >= flag", id, surface)
			}
		}
	}
	for _, rule := range p.Rules {
		if rule.Signal != "" {
			if _, ok := p.Signals[rule.Signal]; !ok {
				return fmt.Errorf("rule %q refers to unknown signal %q", rule.ID, rule.Signal)
			}
		}
		if rule.AddFinding != "" {
			if _, ok := p.Categories[rule.AddFinding]; !ok {
				return fmt.Errorf("rule %q adds unknown category %q", rule.ID, rule.AddFinding)
			}
		}
	}
	return nil
}

func stringOr(v any, fallback string) string {
	if v == nil {
		return fallback
	}
	return pyStr(v)
}

func mapOf(v any) map[string]any {
	if m, ok := v.(map[string]any); ok {
		return m
	}
	return map[string]any{}
}

func stringsOf(v any) []string {
	items, ok := v.([]any)
	if !ok {
		return nil
	}
	out := make([]string, 0, len(items))
	for _, item := range items {
		out = append(out, pyStr(item))
	}
	return out
}

File: ts/src/policy.ts
/** Loading and querying a policy pack. */

import { readFileSync } from "node:fs";
import type { Surface } from "./types.js";

export interface Thresholds {
  readonly block: number;
  readonly review: number;
  readonly flag: number;
}

export interface CategorySpec {
  readonly name: string;
  readonly description: string;
  readonly refs?: readonly string[];
  readonly surfaces?: readonly Surface[];
  readonly weight?: number;
  readonly base_severity?: number;
  readonly sentinel?: boolean;
  readonly sentinel_instructions?: string;
  readonly thresholds: Readonly<Record<string, Thresholds>>;
  readonly route?: "redact" | "guide" | "crisis_support";
  readonly never_below?: string;
  readonly enabled?: boolean;
}

export interface SignalSpec {
  readonly type: "noul" | "choice" | "score";
  readonly surfaces: readonly Surface[];
  readonly instructions: string;
  readonly criteria?: unknown;
  readonly criteria_ref?: string;
  readonly requires_context?: boolean;
}

export interface RuleSpec {
  readonly id: string;
  readonly when: { readonly signal: string; readonly op: string; readonly value: unknown };
  readonly except_categories?: readonly string[];
  readonly then: {
    readonly upgrade?: number;
    readonly downgrade?: number;
    readonly cap_action?: string;
    readonly floor_action?: string;
    readonly add_finding?: string;
    readonly skip_confidence_gate?: boolean;
  };
  readonly why?: string;
}

export interface PolicyPack {
  readonly id: string;
  readonly version: string;
  readonly name?: string;
  readonly content_note?: string;
  readonly defaults?: Readonly<Record<string, unknown>>;
  readonly categories: Readonly<Record<string, CategorySpec>>;
  readonly signals: Readonly<Record<string, SignalSpec>>;
  readonly rules?: readonly RuleSpec[];
  readonly routes?: Readonly<Record<string, string>>;
  readonly [scale: string]: unknown;
}

/** How a pack asks for a sentinel to be backed by the hazard choice. */
export interface SentinelCorroboration {
  /** The hazard-choice probability below which a sentinel counts as uncorroborated. */
  readonly minChoice: number;
  /** The refusal score from which a reply counts as declining. */
  readonly refusal: number;
  /** The sentinel answer from which even a declining reply is not capped. */
  readonly refusalMaxSentinel: number;
  /** Categories a declining reply can still carry, so they are never capped. */
  readonly refusalExcept: ReadonlySet<string>;
  /**
   * Cap an uncorroborated sentinel below its block band at flag: it is recorded and delivered,
   * rather than held, unless it crosses the block band on its own.
   */
  readonly weakAtMostFlag: boolean;
  /**
   * Resolve an uncorroborated sentinel for a category handled by redaction to review, so it is
   * masked and delivered, unless it reaches this value. Zero is off.
   */
  readonly redactInsteadOfBlockBelow: number;
  /** Categories never weakened: an uncorroborated sentinel keeps never_below and is not capped. */
  readonly weakExcept: ReadonlySet<string>;
}

/** When a low-confidence answer escalates to review. */
export interface ConfidenceGateOptions {
  /** Count only findings and near misses the hazard choice backs. */
  readonly needsCorroboration: boolean;
  /**
   * Intent labels for which the gate does not escalate, as long as the answer's confidence is at
   * least `skipMinConfidence`: below that the intent is itself a guess.
   */
  readonly skipWhenIntent: ReadonlySet<string>;
  readonly skipMinConfidence: number;
}

export interface Category extends CategorySpec {
  readonly id: string;
}

const BUNDLED = "standard-v1";

/**
 * A parsed policy pack.
 *
 * The pack is the hazard taxonomy, the thresholds that turn a probability into an action, and
 * the rules that adjust it. Its descriptions are also the text sent to Jev as question criteria,
 * so editing the pack changes both what the model is asked and how the answer is judged.
 */
export class Policy {
  readonly id: string;
  readonly version: string;
  readonly name: string;
  readonly contentNote: string;
  readonly categories: ReadonlyMap<string, Category>;
  readonly signals: Readonly<Record<string, SignalSpec>>;
  readonly rules: readonly RuleSpec[];
  readonly routes: Readonly<Record<string, string>>;
  readonly raw: PolicyPack;

  constructor(pack: PolicyPack) {
    this.raw = pack;
    this.id = pack.id ?? "custom";
    this.version = pack.version ?? "0";
    this.name = pack.name ?? this.id;
    this.contentNote = pack.content_note ?? "";
    this.signals = pack.signals ?? {};
    this.rules = pack.rules ?? [];
    this.routes = pack.routes ?? {};
    this.categories = new Map(
      Object.entries(pack.categories ?? {}).map(([id, spec]) => [id, { ...spec, id }]),
    );
    validate(this);
  }

  /** Load the bundled pack, a pack object, or a pack read from a JSON file. */
  static load(source?: string | PolicyPack | Policy): Policy {
    if (source instanceof Policy) return source;
    if (source === undefined) return Policy.bundled();
    if (typeof source !== "string") return new Policy(source);
    if (!source.includes("/") && !source.endsWith(".json")) return Policy.bundled(source);
    return new Policy(JSON.parse(readFileSync(source, "utf8")) as PolicyPack);
  }

  static bundled(name: string = BUNDLED): Policy {
    const url = new URL(`./policies/${name}.json`, import.meta.url);
    return new Policy(JSON.parse(readFileSync(url, "utf8")) as PolicyPack);
  }

  /** Enabled categories that apply to a surface, most serious first. */
  forSurface(surface: Surface): Category[] {
    return [...this.categories.values()]
      .filter((c) => (c.enabled ?? true) && (c.surfaces ?? []).includes(surface))
      .sort(
        (a, b) =>
          (b.weight ?? 0.5) - (a.weight ?? 0.5) ||
          (b.base_severity ?? 2) - (a.base_severity ?? 2) ||
          a.id.localeCompare(b.id),
      );
  }

  /**
   * Categories that get a dedicated yes/no question on this surface.
   *
   * A choice question picks one label. Content can carry more than one hazard at once, and the
   * ones where a miss is unacceptable get their own independent question.
   */
  sentinels(surface: Surface): Category[] {
    return this.forSurface(surface).filter((c) => c.sentinel && c.sentinel_instructions);
  }

  signalsFor(surface: Surface, hasContext = false): Record<string, SignalSpec> {
    const out: Record<string, SignalSpec> = {};
    for (const [name, spec] of Object.entries(this.signals)) {
      if (!spec.surfaces.includes(surface)) continue;
      if (spec.requires_context && !hasContext) continue;
      out[name] = spec;
    }
    return out;
  }

  /** Resolve a signal's criteria, following `criteria_ref` into the pack's scales. */
  criteriaFor(spec: SignalSpec): unknown {
    if (spec.criteria !== undefined) return spec.criteria;
    return spec.criteria_ref ? this.raw[spec.criteria_ref] : undefined;
  }

  thresholds(category: Category, surface: Surface): Thresholds | undefined {
    return category.thresholds[surface] ?? category.thresholds["default"];
  }

  minConfidence(): number {
    return Number(this.raw.defaults?.["min_confidence"] ?? 0.65);
  }

  onLowConfidence(): string {
    return String(this.raw.defaults?.["on_low_confidence"] ?? "escalate");
  }

  /**
   * Whether an unreachable Jev blocks on this surface.
   *
   * Set per surface, because the two are not the same risk: the input check sits in front of a
   * model that has its own safety, so failing open there degrades gracefully, while the output
   * check is the last line and has nothing behind it.
   */
  failClosed(surface?: Surface): boolean {
    const setting = this.raw.defaults?.["on_error"] ?? "fail_closed";
    if (setting && typeof setting === "object") {
      return String((setting as Record<string, unknown>)[surface ?? ""] ?? "fail_closed") === "fail_closed";
    }
    return String(setting) === "fail_closed";
  }

  /**
   * The pack's `defaults.sentinel_corroboration`, or undefined when the pack does not set it.
   *
   * It is how a pack asks for a sentinel to be backed by the hazard choice before it can drive the
   * strongest actions on its own.
   */
  sentinelCorroboration(): SentinelCorroboration | undefined {
    const raw = this.raw.defaults?.["sentinel_corroboration"];
    if (!isRecord(raw)) return undefined;
    return {
      minChoice: numberOr(raw["min_choice"], 0.02),
      refusal: numberOr(raw["refusal"], 0.8),
      refusalMaxSentinel: numberOr(raw["refusal_max_sentinel"], 0.5),
      refusalExcept: new Set(stringsOf(raw["refusal_except"])),
      weakAtMostFlag: truthy(raw["weak_at_most_flag"]),
      redactInsteadOfBlockBelow: numberOr(raw["redact_instead_of_block_below"], 0),
      weakExcept: new Set(stringsOf(raw["weak_except"])),
    };
  }

  /** The pack's `defaults.confidence_gate` options. */
  confidenceGate(): ConfidenceGateOptions {
    const setting = this.raw.defaults?.["confidence_gate"];
    const raw = isRecord(setting) ? setting : {};
    return {
      needsCorroboration: truthy(raw["needs_corroboration"]),
      skipWhenIntent: new Set(stringsOf(raw["skip_when_intent"])),
      skipMinConfidence: numberOr(raw["skip_min_confidence"], 0.5),
    };
  }

  errorAction(): Action {
    return (this.raw.defaults?.["error_action"] ?? "review") as Action;
  }

  get policyId(): string {
    return `${this.id}@${this.version}`;
  }
}

import type { Action } from "./types.js";

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function numberOr(value: unknown, fallback: number): number {
  if (typeof value === "number") return value;
  if (typeof value === "boolean") return value ? 1 : 0;
  if (typeof value === "string" && value.trim() !== "") {
    const parsed = Number(value);
    if (!Number.isNaN(parsed)) return parsed;
  }
  return fallback;
}

function stringsOf(value: unknown): string[] {
  return Array.isArray(value) ? value.map((item) => String(item)) : [];
}

/** Truthiness the way the Python and Go packages read a pack setting. */
function truthy(value: unknown): boolean {
  if (value === null || value === undefined) return false;
  if (typeof value === "boolean") return value;
  if (typeof value === "string") return value !== "";
  if (Array.isArray(value)) return value.length > 0;
  if (typeof value === "object") return Object.keys(value).length > 0;
  return typeof value !== "number" || value !== 0;
}

function validate(policy: Policy): void {
  if (policy.categories.size === 0) throw new Error("policy pack has no categories");
  for (const [id, category] of policy.categories) {
    const banded = Object.entries(category.thresholds ?? {});
    if (banded.length === 0) throw new Error(`category "${id}" has no thresholds`);
    for (const [surface, bands] of banded) {
      for (const band of ["block", "review", "flag"] as const) {
        if (typeof bands[band] !== "number") {
          throw new Error(`category "${id}" thresholds[${surface}] is missing "${band}"`);
        }
      }
      if (!(bands.block >= bands.review && bands.review >= bands.flag)) {
        throw new Error(
          `category "${id}" thresholds[${surface}] are not ordered block >= review >= flag`,
        );
      }
    }
  }
  for (const rule of policy.rules) {
    if (rule.when?.signal && !(rule.when.signal in policy.signals)) {
      throw new Error(`rule "${rule.id}" refers to unknown signal "${rule.when.signal}"`);
    }
    const added = rule.then?.add_finding;
    if (added && !policy.categories.has(added)) {
      throw new Error(`rule "${rule.id}" adds unknown category "${added}"`);
    }
  }
}
